lib/dirac_coarse.cpp

```diff
@@ -585,68 +577,56 @@ namespace quda {
-        src[i] = const_cast<ColorSpinorField &>(b[i]).create_alias();
-        sol[i] = x[i].create_alias();
-      }
+      create_alias(src, b);
+      create_alias(sol, x);
       return;
     }
 
-    auto tmp = getFieldTmp(b[0].Even());
+    create_alias(src, x(other_parity));
+    create_alias(sol, x(this_parity));
 
+    auto tmp = getFieldTmp(x.Even());
     // we desire solution to full system
-    for (auto i = 0u; i < b.size(); i++) {
-
-      if (symmetric) {
-        // src = A_ee^-1 (b_e - D_eo A_oo^-1 b_o)
-        src[i] = x[i][other_parity].create_alias();
+    if (symmetric) {
+      // src = A_ee^-1 (b_e - D_eo A_oo^-1 b_o)
 #if 0
-        CloverInv(src[i], b[other_parity], other_parity);
-        DiracCoarse::Dslash(tmp, src[i], this_parity);
-        blas::xpay(b[i][this_parity], -1.0, tmp);
-        CloverInv(src[i], tmp, this_parity);
+      CloverInv(src, b(other_parity), other_parity);
+      DiracCoarse::Dslash(tmp, src, this_parity);
+      blas::xpay(b(this_parity), -1.0, tmp);
+      CloverInv(src, tmp, this_parity);
+#else
+      // src = A_ee^{-1} b_e - (A_ee^{-1} D_eo) A_oo^{-1} b_o
+      CloverInv(src, b(other_parity), other_parity);
+      Dslash(tmp, src, this_parity);
+      CloverInv(src, b(this_parity), this_parity);
+      blas::axpy(-1.0, tmp, src);
 #endif
-        // src = A_ee^{-1} b_e - (A_ee^{-1} D_eo) A_oo^{-1} b_o
-        CloverInv(src[i], b[i][other_parity], other_parity);
-        Dslash(tmp, src[i], this_parity);
-        CloverInv(src[i], b[i][this_parity], this_parity);
-        blas::axpy(-1.0, tmp, src[i]);
-
-        sol[i] = x[i][this_parity].create_alias();
-      } else {
-        // src = b_e - D_eo A_oo^-1 b_o
-        src[i] = x[i][other_parity].create_alias();
-        CloverInv(tmp, b[i][other_parity], other_parity);
-        DiracCoarse::Dslash(src[i], tmp, this_parity);
-        blas::xpay(b[i][this_parity], -1.0, src[i]);
-        sol[i] = x[i][this_parity].create_alias();
-      }
+    } else {
+      // src = b_e - D_eo A_oo^-1 b_o
+      CloverInv(tmp, b(other_parity), other_parity);
+      DiracCoarse::Dslash(src, tmp, this_parity);
+      blas::xpay(b(this_parity), -1.0, src);
     }
   }
 
   void DiracCoarsePC::reconstruct(cvector_ref<ColorSpinorField> &x, cvector_ref<const ColorSpinorField> &b,
                                   const QudaSolutionType solType) const
   {
-    if (solType == QUDA_MATPC_SOLUTION || solType == QUDA_MATPCDAG_MATPC_SOLUTION) {
-      return;
-    }
-
-    auto tmp = getFieldTmp(b[0].Even());
-
-    for (auto i = 0u; i < b.size(); i++) {
-      checkFullSpinor(x[i], b[i]);
-
-#if 0
-      // x_o = A_oo^-1 (b_o - D_oe x_e)
-      DiracCoarse::Dslash(tmp, x.Even(), QUDA_ODD_PARITY);
-      blas::xpay(b.Odd(), -1.0, tmp);
-      CloverInv(x.Odd(), tmp, QUDA_ODD_PARITY);
+    if (solType == QUDA_MATPC_SOLUTION || solType == QUDA_MATPCDAG_MATPC_SOLUTION) return;
+
+    checkFullSpinor(x, b);
+    auto tmp = getFieldTmp(x.Even());
+#if 1
```

Comment:
This needs to be switched back to `#if 0`, I'm guessing this was for some intermediate debugging purposes. The origin of this was when we were (more) memory constrained with HISQ MG and we wanted to keep the non-preconditioned links fully on the host in the managed-memory mode.

Reply:
e89be7d340ad72d37cec7a83c4d1fc25d8f85a75